### PR TITLE
Convert setup.py to pyproject.toml

### LINKED ISSUE #1563: Move project configuration from setup.py to pyproject.toml (PEP 621)
See [here](https://peps.python.org/pep-0621/) and [here](https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html).

### PR DESCRIPTION
**Pull request by Tbruno25**
Closes #1563

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -1,19 +1,141 @@
...
+    "I", # isort
 
-# Assume Python 3.7.
-target-version = "py37"
+]
 
 [tool.ruff.isort]
 known-first-party = ["can"]
+
+[build-system]
```
Comment (zariiii9003) [inline pyproject.toml:139]: This is usually at the top of the file

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -84,7 +84,7 @@ jobs:
     - name: Install dependencies
       run: |
         python -m pip install --upgrade pip
-        pip install -r requirements-lint.txt
+        pip install .[lint]
```
Comment (zariiii9003) [inline .github/workflows/ci.yml:87]:
```suggestion
        pip install -e .[lint]
```

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -125,7 +125,7 @@ jobs:
...
```
Comment (zariiii9003) [inline .github/workflows/ci.yml:128]:
```suggestion
        pip install -e .[lint]
```

File: .github/workflows/format-code.yml [hunk 4]
```diff
@@ -17,7 +17,7 @@ jobs:
     - name: Install dependencies
       run: |
         python -m pip install --upgrade pip
-        pip install -r requirements-lint.txt
+        pip install .[lint]
```
Comment (zariiii9003) [inline .github/workflows/format-code.yml:20]:
```suggestion
        pip install -e .[lint]
```

File: pyproject.toml [hunk 5]
```diff
@@ -1,19 +1,141 @@
...
+    "License :: OSI Approved :: GNU Lesser General Public License v3 (LGPLv3)",
+    "Natural Language :: English",
+    "Natural Language :: English",
+    "Operating System :: MacOS",
+    "Operating System :: Microsoft :: Windows",
+    "Operating System :: POSIX :: Linux",
+    "Programming Language :: Python",
+    "Programming Language :: Python :: 3.10",
+    "Programming Language :: Python :: 3.11",
+    "Programming Language :: Python :: 3.7",
```
Comment (zariiii9003) [inline pyproject.toml:32]: the python versions could be sorted here 😄

File: pyproject.toml [hunk 6]
```diff
@@ -1,19 +1,141 @@
...
+dependencies = [
+    "wrapt~=1.10",
+    "typing_extensions>=3.10.0.0",
+    "pywin32>=305; platform_system == 'Windows' and platform_python_implementation == 'CPython'",
+    "msgpack~=1.0.0; platform_system != 'Windows'",
     "setuptools >= 40.8",
-    "wheel",
 ]
-build-backend = "setuptools.build_meta"
+requires-python = ">=3.7, <=4.0"
```
Comment (zariiii9003) [inline pyproject.toml:13]:
```suggestion
requires-python = ">=3.7"
```

File: pyproject.toml [hunk 7]
```diff
@@ -1,19 +1,141 @@
...
 
-# Assume Python 3.7.
-target-version = "py37"
+]
 
 [tool.ruff.isort]
 known-first-party = ["can"]
+
+[build-system]
+requires = ["packaging", "setuptools >= 40.8", "wheel"]
```
Comment (zariiii9003) [inline pyproject.toml:140]: `packaging` is a runtime dependency, not a build dependency. `setuptools` must be >=61 for pyproject.toml support, but i would just set the latest version as the new minimum. `wheel` is not necessary anymore.

**Review by zariiii9003 — CHANGES_REQUESTED** [review]

**Comment by Tbruno25** [comment]
Updated to include the suggestions! :rocket:

**Review by zariiii9003 — APPROVED** [review]
Looks good, thanks!

**Pull request merged**